### ISSUE TITLE
Send update pings to AMP Caches

### ISSUE THREAD

**Issue by mjangda** [ampproject/amp-wp#598]
On post update and delete. See ~https://developers.google.com/amp/cache/update-ping~ https://developers.google.com/amp/cache/update-cache

**Comment by westonruter** [comment]
Related: #2382 (Dynamically load list of AMP Caches).

**Comment by westonruter** [comment]
We need to get more clarity on this API before we dig into this.

**Comment by jono-alderson** [comment]
What needs to happen to move this forward?

I see a _lot_ of problems and complaints out in the wild about cache staleness, the lack of tools to sync/update data, etc.

Anecdotally, it's a _big_ barrier to adoption from a lot of organizations who might otherwise be convinced to make the leap!

**Comment by archon810** [comment]
Yes! I'd say this is the #1 issue we're still facing with AMP daily.

I recently tried to make another effort to revive this task so hopefully we'll get some much-needed clarity regarding caches soon so that @westonruter can proceed with the implementation.

https://github.com/ampproject/amp.dev/issues/3348

https://github.com/ampproject/amphtml/issues/24326

cc @morsssss

**Comment by westonruter** [comment]
Yes, we can't really move forward with some cache purging logic in the AMP plugin until there is more clarity on purging the AMP Cache(s) generally.

**Comment by laras126** [comment]
Hi there @westonruter! My team is looking to build an admin interface for AMP Cache purging by URL. Our implementation plan is as follows:

* Serving public key
   * Generate a common keypair per the documentation
   * Hook in to request to intercept the .well-known/amphtml/apikey.pub request
   * Set correct content-type header, and any other headers mentioned in the documentation
   * Output the public key
* Interface for purging
   * Create a submenu page under the AMP plugin
   * In lieu of saving the submitted value, queue a purge-request cron job (for immediate execution, to make it async) and discard value
* Cron purge job
   * Retrieve and cache list of AMP caches as provided by AMP project
   * Generate signed URL and submit to each AMP cache

I encountered this issue for sending update pings, and wanted to check – would the above scope be useful as a contribution to the AMP WP plugin? Or are there still issues with purging AMP Cache that need to be clarified before this can be pursued (per above comment from Jan '20)?

cc @ethitter

**Comment by westonruter** [comment]
Hi ~@laras~ _@laras126_! Good to hear from you. 

The implementation plan seems sound. My main concern is that the `update-cache` API doesn't appear to be work reliably per https://github.com/ampproject/amphtml/issues/24326 or at least not quickly. But I don't have much experience with it, to be honest. To get more clarity on that I'd recommend following up with the AMP core team in that issue.

What I suggest is that you first implement this in a standalone extension plugin for the AMP plugin and verify that it works as expected. By being in a separate “feature plugin” to start with, there won't be issues with the PR going stale or being blocked by a plugin release. Once the feature plugin is working then it could be merged into the AMP plugin directly.

**Comment by LaraS** [comment]
Hi @laras126, I think this tag was for you.
